state-sync/inter-component/event-notifications/src/lib.rs

```diff
@@ -0,0 +1,458 @@
+        let account_state_blob = account_state_blob.ok_or_else(|| {
+            Error::UnexpectedErrorEncountered("Missing account state blob!".into())
+        })?;
+
+        // Fetch the new epoch from storage
+        let epoch = AccountState::try_from(&account_state_blob)
+            .and_then(|state| {
+                Ok(state
+                    .get_configuration_resource()?
+                    .ok_or_else(|| {
+                        Error::UnexpectedErrorEncountered(
+                            "Configuration resource does not exist!".into(),
+                        )
+                    })?
+                    .epoch())
+            })
+            .map_err(|error| {
+                Error::UnexpectedErrorEncountered(format!(
+                    "Failed to fetch configuration resource! Error: {:?}",
+                    error
+                ))
+            })?;
+
+        // Return the config payload
+        Ok(OnChainConfigPayload::new(
+            epoch,
+            Arc::new(
+                ON_CHAIN_CONFIG_REGISTRY
+                    .iter()
+                    .cloned()
+                    .zip_eq(on_chain_configs)
+                    .collect(),
+            ),
+        ))
+    }
+}
+
+#[async_trait]
+impl EventNotificationSender for EventSubscriptionService {
+    async fn notify_events(
+        &mut self,
+        version: Version,
+        events: Vec<ContractEvent>,
+    ) -> Result<(), Error> {
+        if events.is_empty() {
+            return Ok(()); // No events!
+        }
+
+        // Notify event subscribers and check if a reconfiguration event was processed
+        let reconfig_event_processed = self.notify_event_subscribers(version, events)?;
+
+        // If a reconfiguration event was found, also notify the reconfig subscribers
+        // of the new configuration values.
+        if reconfig_event_processed {
+            self.force_notify_reconfigurations(version).await
+        } else {
+            Ok(())
+        }
+    }
+
+    async fn force_notify_reconfigurations(&mut self, version: Version) -> Result<(), Error> {
+        if self.reconfig_subscriptions.is_empty() {
+            return Ok(()); // No reconfiguration subscribers!
+        }
+
+        let new_configs = self.read_on_chain_configs(version)?;
+        for (_, reconfig_subscription) in self.reconfig_subscriptions.iter_mut() {
+            reconfig_subscription.notify_subscriber_of_configs(version, new_configs.clone())?;
+        }
+
+        Ok(())
+    }
+}
+
+/// A unique ID used to identify each subscription.
+type SubscriptionId = u64;
+
+/// A single event subscription, holding the subscription identifier, channel to
+/// send the corresponding notifications and a buffer to hold pending events.
+#[derive(Debug)]
+struct EventSubscription {
+    pub subscription_id: SubscriptionId,
+    pub event_buffer: Vec<ContractEvent>,
+    pub notification_sender: channel::diem_channel::Sender<(), EventNotification>,
```

Comment:
Should we use a diem_channel if the key is always `()`?  Though, it's kinda weird we have too many types of channels

Reply:
Yeah, this is another thing that confused me and I think we should fix it. But, I've tried not to change too much in this PR (and just keep whatever is currently being used) to keep the PR line count down. But, this is something I'll iterate on.